File: core/plugins/scrapers/btdig.py
"""BTDigg scraper — DHT search engine."""

import json
from urllib.request import Request, urlopen
from urllib.parse import urlencode, quote

from core.plugins.base import TorrentResult


class BTDiggScraper:
    """Search torrents via BTDigg DHT search engine (public, no auth)."""

    name = "btdig"
    category = "public-tracker"
    enabled = True
    base_url = "https://api.btdig.com"

    def search(self, media_type: str, item_id: str, config: dict,
               query: str | None = None, limit: int = 50) -> list[TorrentResult]:
        search_query = query or item_id
        params = {
            "q": search_query,
            "output": "json",
        }
        url = f"{self.base_url}/search?{urlencode(params)}"
        req = Request(url)
        req.add_header("User-Agent", "StreamSyncr/2.0")

        try:
            with urlopen(req, timeout=15) as resp:
                data = json.loads(resp.read())
        except Exception:
            return []

        results = []
        for item in data.get("torrents", [])[:limit]:
            info_hash = item.get("info_hash", "").upper()
            magnet = f"magnet:?xt=urn:btih:{info_hash}&dn={quote(item.get('title', ''))}"
            results.append(TorrentResult(
                title=item.get("title", "Unknown"),
                info_hash=info_hash,
                size=item.get("size", 0),
                seeders=item.get("seeders", 0),
                tracker="BTDigg",
                magnet=magnet,
            ))

        return results

File: nuvio/providers/btdig.js
/**
 * StreamSyncr BTDigg Provider for Nuvio
 * 
 * DHT search engine — free, no auth required.
 */

const BASE_URL = "http://localhost:7800";

async function getStreams(id, type, title, year, season, episode) {
  try {
    const mediaType = type === "tv" ? "series" : type;
    const imdbId = id || "";
    
    const url = `${BASE_URL}/api/v1/streams/${mediaType}/${imdbId}?config_token=${getConfigToken()}`;
    
    const response = await fetch(url, {
      method: "GET",
      headers: {
        "Content-Type": "application/json",
        "User-Agent": "Nuvio-StreamSyncr/2.0",
      },
    });
    
    if (!response.ok) return [];
    
    const data = await response.json();
    return (data.streams || []).map((stream) => ({
      name: stream.name || "StreamSyncr",
      title: stream.title || "Unknown",
      url: stream.url,
      behaviorHints: {
        bingeGroup: `streamsyncr-btdig-${id}`,
      },
    }));
  } catch (error) {
    console.error("StreamSyncr BTDigg error:", error);
    return [];
  }
}

function getConfigToken() {
  return "";
}

if (typeof globalThis !== "undefined") {
  globalThis.getStreams = getStreams;
}
